src/library/scala/collection/Map.scala

```diff
@@ -29,25 +29,14 @@ trait Map[K, +V]
 
   def canEqual(that: Any): Boolean = true
 
-  override def equals(o: Any): Boolean = o match {
-    case that: Map[K, _] =>
-      (this eq that) ||
-      (that canEqual this) &&
-      (this.size == that.size) && {
-        try {
-          val checker = new AbstractFunction1[(K, V),Boolean] {
-            override def apply(kv: (K,V)): Boolean = {
-              that.getOrElse(kv._1, Map.DefaultSentinel) == kv._2
-            }
-          }
-          this forall checker
-        } catch {
-          case _: ClassCastException => false
-        }
-      }
-    case _ =>
-      false
-  }
+  override def equals(o: Any): Boolean =
+    (this eq o.asInstanceOf[AnyRef]) || (o match {
+      case map: Map[K, _] if map.canEqual(this) =>
+        (this.size == map.size) &&
+          this.forall(kv => map.getOrElse(kv._1, Map.DefaultSentinel()) == kv._2)
```

Comment:
There's actually a problem here: what if `map` does *not* contain a binding for `kv._1`, but `kv._2.equals(DefaultSentinel())` is `true` (because of its implementation of `equals`)? Do we need to worry about that?